examples/neutra.py

```diff
@@ -11,32 +14,30 @@
        (https://arxiv.org/abs/1903.03704)
 """
 
-
 import argparse
 from functools import partial
-import os
 
 from matplotlib.gridspec import GridSpec
 import matplotlib.pyplot as plt
 import seaborn as sns
 
 from jax import lax, random, vmap
 import jax.numpy as np
-from jax.scipy.special import logsumexp
 from jax.tree_util import tree_map
 
 import numpyro
 from numpyro import optim
-from numpyro.contrib.autoguide import AutoContinuousELBO, AutoIAFNormal
+from numpyro.contrib.autoguide import AutoContinuousELBO, AutoBNAFNormal
 from numpyro.diagnostics import print_summary
 import numpyro.distributions as dist
 from numpyro.distributions import constraints
 from numpyro.infer import MCMC, NUTS, SVI
 from numpyro.infer.util import initialize_model, transformed_potential_energy
 
-# TODO: remove when the issue https://github.com/google/jax/issues/939 is fixed upstream
-# The behaviour when training guide under fast math mode is unstable.
-os.environ["XLA_FLAGS"] = "--xla_cpu_enable_fast_math=false"
+
+# XXX: upstream logsumexp throws NaN under fast-math mode + MCMC's progress_bar=True
+def logsumexp(x, axis=0):
+    return np.log(np.sum(np.exp(x), axis=axis))
```

Comment:
It took me a day to debug the issue: NeuTra throws nan when using fast-math mode and enable progress_bar. We have noticed this issue previously: https://github.com/pyro-ppl/numpyro/pull/280

It is pretty hard to isolate the issue: I trained AutoBNAF with fast_math disabled, store params in a file, then enable fast-math, load params, create transform_potential_fn, run MCMC, get NaN... NaN also happens for HMC with only 1 leapfrog step. However, when I run the utility velocity_verlet alone (out of MCMC scope), NaN doesn't happen. All I can tell (after diving deeper into the source code): somehow transform_pe has NaN grad under fast-math mode.

Reply:
Interestingly, using `np.log(np.sum(np.exp(x - np.max(x)), axis=axis)) + np.max(x)` here also throws NaN. I have no idea why... I believe 0 x inf is happening in grad somewhere here. If we face this NaN, it is better to disable fast-math mode, until JAX fast-math is more stable.

Reply:
That's some serious sleuthing! Any ideas on what kind of values does it throw `nan`?

Reply:
The input to DualMoon log_prob is just a usual value ~ `np.array([-4., 0.])`. Then, the input to `logsumexp` will be around `np.array([-50., -5.])`. Grad at this value will be `(3e-20, 1)`.

Reply:
I thought this was the function throwing nan, but it seems that the nan is being thrown somewhere else even though the logsumexp function gives finite gradients.

Reply:
This is a bit mysterious. Given the `params` which throw `nan` in grad (I know `grad` is `nan` by changing transition prob logic to capture NaN grad in hmc_state), `value_and_grad(potential_fn)(params)` gives finite results when I run it outside MCMC.